File: CS50X/pset02/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
int calculate_coleman_liau_index(float l, float w, float s);

int main(void)
{
    // Prompt user for a text
    string s = get_string("Text: ");

    // Declare variable to save 's' length
    int l = strlen(s);

    // Count the number of letters, words and sentences from text
    float letters = count_letters(s);
    float words = count_words(s);
    float sentences = count_sentences(s);

    // Calculate Coleman-Liau index
    int index = calculate_coleman_liau_index(letters, words, sentences);

    // Print the text grade level
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

int count_letters(string text)
{
    int c = 0;
    int length = strlen(text);

    for (int i = 0; i < length; i++)
    {
        if (isalpha(text[i]))
        {
            c++;
        }
    }
    return c;
}

int count_words(string text)
{
    int c = 0;
    int length = strlen(text);

    for (int i = 0; i < length; i++)
    {
        if (isspace(text[i]))
        {
            c++;
        }
    }
    return c + 1;
}

int count_sentences(string text)
{
    int c = 0;
    int length = strlen(text);

    for (int i = 0; i < length; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            c++;
        }
    }
    return c;
}

int calculate_coleman_liau_index(float l, float w, float s)
{
    float L = ((l / w) * 100);
    float S = ((s / w) * 100);
    int index = round(0.0588 * L - 0.296 * S - 15.8);

    return index;
}

File: CS50X/pset06/sentimental-readability/readability.py
def main():
    # Prompt user for a text
    text = input("Text: ")

    # Count the letters of the text
    letters = count_letters(text)

    # Count the words of the text
    words = count_words(text)

    # Count the sentences of the text
    sentences = count_sentences(text)

    # Calculated the Coleman-Liau index
    index = calculate_coleman_index(letters, words, sentences)

    # Print the grade of the text
    print_grade(index)


def count_letters(text):

    letters = 0

    # Iterate through the text and count the letters
    for caracter in text:
        if caracter.isalpha():
            letters += 1
    return letters


def count_words(text):

    words = 0

    # Iterate through the text and count the words
    for caracter in text:
        if caracter.isspace():
            words += 1
    return words + 1


def count_sentences(text):

    sentences = 0

    # Iterate through the text and count the sentences
    for caracter in text:
        if caracter in ['.', '!', '?']:
            sentences += 1
    return sentences


def calculate_coleman_index(letters, words, sentences):

    # Calculate the Coleman-Liau index
    L = (letters / words) * 100
    S = (sentences / words) * 100

    return round(0.0588 * L - 0.296 * S - 15.8)


def print_grade(index):

    # Print the grade of the text based in the Coleman-Liau index
    if index < 1:
        print("Before Grade 1")
    elif index >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {index}")


main()
